proxy/src/coco/source.rs

```diff
@@ -0,0 +1,452 @@
+            )
+        };
+
+        Self {
+            sha1: commit.id,
+            author: Person {
+                name: commit.author.name.clone(),
+                email: commit.author.email.clone(),
+                avatar: avatar(&commit.author.email),
+            },
+            summary: commit.summary.clone(),
+            message: commit.message.clone(),
+            committer: Person {
+                name: commit.committer.name.clone(),
+                email: commit.committer.email.clone(),
+                avatar: avatar(&commit.committer.email),
+            },
+            committer_time: commit.author.time,
+        }
+    }
+}
+
+/// Git object types.
+///
+/// `shafiul.github.io/gitbook/1_the_git_object_model.html`
+#[derive(Debug, Eq, Ord, PartialOrd, PartialEq)]
+pub enum ObjectType {
+    /// References a list of other trees and blobs.
+    Tree,
+    /// Used to store file data.
+    Blob,
+}
+
+/// Set of extra information we carry for blob and tree objects returned from the API.
+pub struct Info {
+    /// Name part of an object.
+    pub name: String,
+    /// The type of the object.
+    pub object_type: ObjectType,
+    /// The last commmit that touched this object.
+    pub last_commit: Option<Commit>,
+}
+
+/// File data abstraction.
+pub struct Blob {
+    /// Actual content of the file, if the content is ASCII.
+    pub content: BlobContent,
+    /// Extra info for the file.
+    pub info: Info,
+    /// Absolute path to the object from the root of the repo.
+    pub path: String,
+}
+
+impl Blob {
+    /// Indicates if the content of the [`Blob`] is binary.
+    #[must_use]
+    pub fn is_binary(&self) -> bool {
+        self.content == BlobContent::Binary
+    }
+}
+
+/// Variants of blob content.
+#[derive(PartialEq)]
+pub enum BlobContent {
+    /// Content is ASCII and can be passed as a string.
+    Ascii(String),
+    /// Content is binary and needs special treatment.
+    Binary,
+}
+
+/// Result of a directory listing, carries other trees and blobs.
+pub struct Tree {
+    /// Absolute path to the tree object from the repo root.
+    pub path: String,
+    /// Entries listed in that tree result.
+    pub entries: Vec<TreeEntry>,
+    /// Extra info for the tree object.
+    pub info: Info,
+}
+
+/// Entry in a Tree result.
+pub struct TreeEntry {
+    /// Extra info for the entry.
+    pub info: Info,
+    /// Absolute path to the object from the root of the repo.
+    pub path: String,
+}
+
+/// Returns the [`Blob`] for a file at `revision` under `path`.
+///
+/// # Errors
+///
+/// Will return [`error::Error`] if the project doesn't exist or a surf interaction fails.
+pub async fn blob<C: coco::Client>(
+    client: &C,
+    project_urn: String,
+    revision: Option<String>,
+    maybe_path: Option<String>,
+) -> Result<Blob, error::Error> {
+    let (mut repo, project) = client.get_repo(project_urn).await?;
+    let locked = repo.locked();
+    let mut browser = locked.browser()?;
```

Comment:
It annoyed me that I couldn't go a `get_browser` because of lifetimes. Maybe we could do something like `with_browser` in `Client` that allows us to scope the `Browser` and apply a call back?